### PR TITLE
ENH: Performance improvements, bug-fix, and `iter_regions` method

### LINKED ISSUE #929: slice_from_regex is broken on non-boolean vectors
```
In [1]: from skbio import Sequence

In [2]: import numpy as np

In [3]: a = np.arange(5)

In [4]: s = Sequence("ABCBCACBCBCA")

In [5]: s[s.slices_from_regex("(CB)", ignore=a)]
Out[5]: Sequence('BCBC', length=4)
```

### PR DESCRIPTION
**Pull request by ebolyen**
fixes #929

### CODE REVIEW

**Comment by ebolyen** [comment]
API: changed names of several methods and property
    
 *  `sequence` -> `values`
 * `slices_from_regex` -> `find_with_regex`
 *  `kmers` -> `iter_kmers`
 *  `iter_regions` -> `iter_contiguous` + added `ignore` kwarg

File: skbio/sequence/_sequence.py [hunk 1]
```diff
@@ -1242,6 +1240,62 @@ def slices_from_regex(self, regex, ignore=None):
...
+            skipped instead of included. Default is False.
+
+        Returns
+        -------
+        generator
+            Yields subsequences as indicated by `included`.
+
+        Notes
+        -----
+        If slices provide adjacent ranges, then they will be considered the
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: Can you add a doctest example demonstrating this case?

Comment (ebolyen) [inline skbio/sequence/_sequence.py]: That's what the `find_motifs` example shows, but maybe it isn't obvious enough.

File: skbio/sequence/_sequence.py [hunk 2]
```diff
@@ -1343,6 +1397,45 @@ def _to(self, **kwargs):
...
+            bool_mode = False
+            int_mode = False
+            for s in sliceable:
+                if isinstance(s, (bool, np.bool_)):
+                    bool_mode = True
+                elif isinstance(s, (slice, int, np.signedinteger)) or (
+                        hasattr(s, 'dtype') and s.dtype != np.bool):
+                    int_mode = True
+                else:
+                    raise ValueError("Invalid type in iterable: %s" %
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: I think a `TypeError` makes more sense here.

Comment (jairideout) [inline skbio/sequence/_sequence.py]: Can you note the acceptable types in the error message?

File: skbio/sequence/_sequence.py [hunk 3]
```diff
@@ -1343,6 +1397,45 @@ def _to(self, **kwargs):
...
+                                     s.__class__.__name__)
+            if bool_mode and int_mode:
+                raise ValueError("Cannot provide iterable of both bool and"
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: use `TypeError`

File: skbio/sequence/_sequence.py [hunk 4]
```diff
@@ -1343,6 +1397,45 @@ def _to(self, **kwargs):
...
+                                 " int.")
+            sliceable = np.r_[sliceable]
+
+        if sliceable.dtype == np.bool:
+            if sliceable.size != len(self):
+                raise ValueError("Boolean array does not match length of"
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: Can you add the provided and expected lengths in the error message?

File: skbio/sequence/tests/test_sequence.py [hunk 5]
```diff
@@ -338,17 +339,17 @@ def test_sequence_property(self):
         seq = Sequence('ACGT')
 
         # should get back a numpy.ndarray of '|S1' dtype
-        self.assertIsInstance(seq.sequence, np.ndarray)
-        self.assertEqual(seq.sequence.dtype, '|S1')
-        npt.assert_equal(seq.sequence, np.array('ACGT', dtype='c'))
+        self.assertIsInstance(seq.values, np.ndarray)
```
Comment (jairideout) [inline skbio/sequence/tests/test_sequence.py]: rename `test_sequence_property` to `test_values_property`

File: skbio/sequence/_sequence.py (hunk 1 above)
Comment (jairideout) [inline skbio/sequence/_sequence.py]: Ah I missed that -- I think it'd help to either add a description to that example noting the adjacent behavior or add a new example with adjacent slices.

**Comment by jairideout** [comment]
@ebolyen looks good, just a few minor comments. Thanks!

File: skbio/sequence/_sequence.py (hunk 1 above)
Comment (ebolyen) [inline skbio/sequence/_sequence.py]: fixed

File: skbio/sequence/tests/test_sequence.py (hunk 5 above)
Comment (ebolyen) [inline skbio/sequence/tests/test_sequence.py]: fixed

File: skbio/sequence/_sequence.py [hunk 6]
```diff
@@ -1242,6 +1240,74 @@ def slices_from_regex(self, regex, ignore=None):
...
+        ...     ungapped_subsequence
+        DNA('AAA', length=3)
+        DNA('TT', length=2)
+        DNA('CCCC', length=4)
+
+        Note how the last potential subsequence was skipped because it would
+        have been smaller than our `min_length` which was set to 2.
+
+        We can also use `iter_contiguous` on a generator of slices as is
+        produces by `find_motifs` (and `find_with_regex`).
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: produces -> produced

**Comment by jairideout** [comment]
New commit introduced a typo, once that's fixed and tests pass this is ready for merge.

**Comment by ebolyen** [comment]
Thanks @jairideout!

**Pull request merged**